commit 4143179218960a70d821a425e3c23ce44aa93dee
Author: Lukas Czerner <lczerner@redhat.com>
Date:   Wed Feb 23 12:42:32 2011 -0500

    ext4: check if device support discard in FITRIM ioctl
    
    For a device that does not support discard, the FITRIM ioctl returns
    -EOPNOTSUPP when blkdev_issue_discard() returns this error code, which
    is how the user is informed that the device does not support discard.
    
    If there are no suitable free extents to be trimmed, then FITRIM will
    return success even though the device does not support discard, which
    could confuse the user.  So check explicitly if the device supports
    discard and return an error code at the beginning of the FITRIM ioctl
    processing.
    
    Signed-off-by: Lukas Czerner <lczerner@redhat.com>
    Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>

diff --git a/fs/ext4/ioctl.c b/fs/ext4/ioctl.c
index eb3bc2f..25ba7c7 100644
--- a/fs/ext4/ioctl.c
+++ b/fs/ext4/ioctl.c
@@ -334,12 +334,16 @@ mext_out:
 	case FITRIM:
 	{
 		struct super_block *sb = inode->i_sb;
+		struct request_queue *q = bdev_get_queue(sb->s_bdev);
 		struct fstrim_range range;
 		int ret = 0;
 
 		if (!capable(CAP_SYS_ADMIN))
 			return -EPERM;
 
+		if (!blk_queue_discard(q))
+			return -EOPNOTSUPP;
+
 		if (copy_from_user(&range, (struct fstrim_range *)arg,
 		    sizeof(range)))
 			return -EFAULT;


